source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs

```diff
@@ -421,6 +425,59 @@ void SetupContextForAmazonServiceAccount(string @namespace, string clusterUrl, s
+                    try
+                    {
+                        var region = Regex.Match(clusterUrl, @"(us(-gov)?|ap|ca|cn|eu|sa)-(central|(north|south)?(east|west)?)-\d", RegexOptions.IgnoreCase).Value;
+                        if (string.IsNullOrWhiteSpace(region))
+                        {
+                            log.Error("The EKS cluster Url specified should contain a valid aws region name");
+                        }
+                        
+                        var awsCliCommandRes = ExecuteCommandAndReturnOutput(aws, "--version").FirstOrDefault();
+                        var awsCliVersion = Regex.Match(awsCliCommandRes, @"aws-cli\/(\d+\.)?(\d+\.)?(\*|\d+)").Value.Replace("aws-cli/", string.Empty);
+                        var awsCliSemVer = new SemanticVersion(awsCliVersion);
+                        var eksGetTokenVersion = new SemanticVersion("1.16.156");
+                        var validAwsVersion = awsCliSemVer.CompareTo(eksGetTokenVersion) == 1;
+
+                        if (validAwsVersion)
+                        {
+                            var awsEksTokenCommand = ExecuteCommandAndReturnOutput("aws",
+                                                                                   "eks",
+                                                                                   "get-token",
+                                                                                   $"--cluster-name={clusterName}",
+                                                                                   $"--region={region}")
+                                .FirstOrDefault();
+                            var apiVersion = JObject.Parse(awsEksTokenCommand).SelectToken("apiVersion");
+
+                            ExecuteKubectlCommand("config",
+                                                  "set-credentials",
+                                                  user,
+                                                  "--exec-command=aws",
+                                                  "--exec-arg=eks",
+                                                  "--exec-arg=get-token",
+                                                  $"--exec-arg=--cluster-name={clusterName}",
+                                                  $"--exec-arg=--region={region}",
+                                                  $"--exec-api-version={apiVersion}");
+
+                            return;
+                        }
+
+                        log.Info($"aws cli version: {awsCliSemVer.ToString()} does not support the \"aws eks get-token\" command. Falling back to aws-iam-authenticator. Please update to a version later than 1.16.156");
+
+                    }
+                    catch (Exception e)
+                    {
+                        log.Verbose($"Unable to authenticate to {clusterUrl} using the aws cli. Failed with the error: {e}.");
+                        log.Info("Falling back to aws-iam-authenticator.");
```

Comment:
I understand that this means that if we use `aws eks get-token` to authenticate, but then the authentication fails, we will attempt to use `aws-iam-authenticator` to authenticate again. I'm not sure if this is intended (or am I understanding this correctly). Also note that there might be other stuff than authentication that may throw exceptions in the try block, for example, `new SemanticVersion()` might fail if an invalid version string is passed in; and the `JObject.Parse` method may also throw if it fails to parse. I think we should probably let the exception happen so we have a chance to fix them.